build/pluginctl/main.go

```diff
@@ -80,6 +81,8 @@ func getClient() (*model.Client4, error) {
 		return nil, errors.New("MM_SERVICESETTINGS_SITEURL is not set")
 	}
 
+	siteURL = strings.TrimSuffix(siteURL, "/")
```

Comment:
This is fine for `pluginctl`, but as you hinted in the summary, sounds like its a "bigger problem" that might be better addressed in `model.NewAPIv4Client`. 

Can you confirm that all Go client usage of the API is broken when there's a trailing slash? If so, would it make sense to strip that trailing slash in `model.NewAPIv4Client` in the server code instead?

Reply:
Great point... I actually started digging deeper to figure out specifically why the call to `client.UploadPluginForced` claims it succeeds but really doesn't... 

I'm still digging, but so far, here is what I see:
1. When there is a trailing slash, that call generates  POST request, but the server is logging a GET and returning a 200 (so the `deploy()` routine believes it was successful and moves on to the next call to enable the plugin (which fails). I find an entry like this in my server logs:

```
{"level":"debug","ts":1599004482.4648635,"caller":"web/handlers.go:100","msg":"Received HTTP request","method":"GET","url":"/api/v4/plugins","request_id":"936f59oqg7d57y9hr94hhnhizc","status_code":"200"}
```

- When there is NOT a trailing slash, the call generates a POST request, and the server correctly logs a POST and returns a 201... everything proceeds as it should from here... I find an entry like this in my server logs:

```
{"level":"debug","ts":1599004263.3228014,"caller":"web/handlers.go:100","msg":"Received HTTP request","method":"POST","url":"/api/v4/plugins","request_id":"75rneq55d7ywuboo9b3ia5nrqa","status_code":"201"}
```

It's extremely odd that, when including the trailing slash, the server seems to be mis-handling it (and even logging it incorrectly too!).  At this point, I'm not 100% sold on where the problem lies, but wanted to get an update (haven't had too much time to put into it).  My guess is that there's quite a few things rearing their heads here... 

To your original question though - does all go client usage break... Yes. Calls, for example, to enable and disable the plugin will fail if there is a trailing slash (I'm running MM on a windows server, but that shouldn't matter).  So yes, if you're looking for the "premium" fix, I agree that we should put that fix over in `model.NewAPIv4Client`... I noticed that this project is still referencing an interesting commit w/in the mattermost-server repo... Can you point me into a direction as to where I should even submit that fix against?  I suppose we can close this PR here and open up an issue over there... Thoughts?

Reply:
Thanks for the deep investigation, @shieldsjared! I suspect the behaviour you observed is a result of the server issuing a redirect from `//...` to `/...` without directing the client to resubmit the `POST` against the new endpoint.

My proposal to fix this:
* Open a PR in https://github.com/mattermost/mattermost-server modifying `model.NewAPIv4Client` with effectively the `siteURL = strings.TrimSuffix(siteURL, "/")` we already have. A quick unit test can verify these changes, and then all of our existing unit tests will exercise these APIs programmatically to verify nothing else is amiss.
* Keep this PR open, and once merged, just update the go module dependency to point at the new `master` commit. The `go.mod` looks weird because it's referencing a pseudo-version from the `master` branch vs. a tagged release -- namely commit `ed34468996e6` (from around July 23rd).

Does that make sense? Happy to help however I can :)

Reply:
Hey @shieldsjared - let us know if you have any questions about Jesse's feedback above. We'd be happy to clarify :)

Reply:
Thanks @jasonblais It all makes sense :) Just haven't had a chance to pop over and execute next steps.  I'll get it done shortly, don't want to leave such a simple thing like this lingering. Sorry!

Reply:
@shieldsjared All good! No worries, thanks for the quick reply :)

Reply:
This code is OBE now. I went ahead and removed it from the PR.